### PR TITLE
chore(test): improve locale test coverage

### PR DESCRIPTION
**Pull request by ccwilson-us-ibm-com**
### Related Ticket(s)

Refs #2080

### Description

Improved test coverage of the Analytics API.

### Changelog

**Changed**

- Added tests for Locale service

### CODE REVIEW

**Comment by ccwilson-us-ibm-com** [comment]
@asudoh I'm not importing my dependencies at the top of the file.  I'm using `jest.doMock` to create my mocks in a `beforeEach` clause.  (`doMock` prevents the mock from being auto-hoisted to the top scope of the file.)

This is a specific strategy that I like to employ with my unit testing.  I like to reduce possible test order dependencies by avoiding variables scoped for the whole file and using `beforeEach` and the `this` context to establish all my mocks and test data.  And then each individual test can adjust the test data or the mock behavior for whatever it needs to do.  This way I can be sure that each tests starts with the same initial state that resets itself in `beforeEach`.

If this doesn't match the team convention, I can change things out.

File: packages/services/src/services/Locale/Locale.js [hunk 1]
```diff
@@ -133,7 +133,12 @@ const _getLocaleByLangAttr = () => {
 async function _getLocaleFromDDO() {
   const ddoLocal = await DDOAPI.getAll();
 
-  if (ddoLocal && ddoLocal.page && ddoLocal.page.pageInfo) {
+  if (
+    ddoLocal &&
+    ddoLocal.page &&
+    ddoLocal.page.pageInfo &&
+    ddoLocal.page.pageInfo.ibm
```
[the remark this answers is not in the document]
Comment (ccwilson-us-ibm-com) [reply packages/services/src/services/Locale/Locale.js]: Thanks.  Updated.

**Comment by asudoh** [comment]
@ccwilson-us-ibm-com Thank you very much trying to explain the background of your pattern. Do you want to elaborate this with some actual code examples and explain why the existing pattern makes the test flaky? Especially I imagine that other Jest users may have hit the same problem given how populate Jest is and the pattern is about very basic mocking strategy. Thanks!

**Comment by jeffchew** [comment]
@asudoh @ccwilson-us-ibm-com just checking on this PR as it looks like conversations on this has stopped a few days ago. Seeing that this is a low priority issue, will move to the next release regardless.

**Comment by ccwilson-us-ibm-com** [comment]
Let's say I mock my module in the file scope:
```
jest.mock('../module');
```

But that's not very useful on it's own for a module that is being imported by the file I'm testing, but not by the test file itself.  So I should provide a mock as well:
```
jest.mock('../module', { ...mockObject });
```

And that works fine, as long as every test needs the same behavior from the mock object.  The moment I need to change how that mock behaves on a per test basis, I need to make some changes.  I need to store the mock in a variable that can be referenced in the test scopes:
```
let mockModule = {....};
jest.mock('../module', mockModule);

describe(function() {
  it('test1', function() {
     mockModule.propertyOrSpy = '...';
     ... do your test
  });

  it('test2', function() {
     ... do your test
  });
});
```

We now have a potential test run order dependency.  If test2 depends on the original behavior of the mock, it may fail when test1 runs first and modifies that behavior.  Ok then... Let's use beforeEach then:
```
let mockModule = {....};
jest.mock('../module', mockModule);

describe(function() {
  beforeEach(function() {
     mockModule.propertyOrSpy = ...;
  });

  it('test1', function() {
     mockModule.propertyOrSpy = '...';
     ... do your test
  });

  it('test2', function() {
     ... do your test
  });
});
```

But now I've introduced a bit of extra cognitive effort.  The mock is defined up top in the file scope, but that definition will never be used, because the before each will overwrite that behavior on every test run.  We could just declare the variable in the file scope (`let mockModule;`), but we're only doing that because the `jest.mock` call is also at the file scope.  This next step is almost entirely personal preference, but is rooted in the old coding principals of avoiding global-ish scopes and defining variables/properties at the lowest scope necessary.  We don't have to split the declaration of the variable from the initialization of the variable, by using the `this` context object within the `beforeEach`, but that also requires using the `jest.doMock` function since `jest.mock` automatically scopes up to the file:
```
describe(function() {
  beforeEach(function() {
     this.mockModule = {};
     jest.doMock('../module', this.mockModule);
  });

  it('test1', function() {
     this.mockModule.propertyOrSpy = '...';
     ... do your test
  });

  it('test2', function() {
     ... do your test
  });
});
```

This reduces the amount of vertical space in the file between variable declaration and initialization.  The same `this` context object is passed to each `it` test.  And the whole state of mocks and data initialized in the `beforeEach` is guaranteed to reset itself before each test.  This keeps test order dependencies from creeping in. (Some test frameworks encourage randomizing the test order to help catch this type of order dependency).  And helps me organize my test data and mock objects.

**Comment by asudoh** [comment]
Thanks @ccwilson-us-ibm-com very much for explaining the problem you are tacking, it's 💯 what I wanted as an explanation. And great to see that you are trying to solve the problem space, it (clean-ups before or after test) is something I see lacking in React codebase in several places.

Another goal in this problem space would be having people, especially new people coming to the codebase, easily understand what we do. From that standpoint, I recommend the following:

```javascript
let mockModule = {....};
jest.mock('../module', mockModule);

// Alternative: Keeping a deep-copy of `mockModule`
const origProperty = mockModule.property;

describe(function() {
  beforeEach(function() {
     mockModule.property = origProperty;
  });

  it('test1', function() {
     mockModule.property = jest.fn();
     ... do your test
  });

  it('test2', function() {
     ... do your test
  });
});
```

Another approach we can take, which seems to be applicable to several (or all?) mocks you added in this change, is leveraging `jest.mock()`'s "turning every function property to mocks" feature and Jest's "auto-resetting all spies/stubs after each test" feature. This allows us to define stubs without having to replace properties or having to manually reset the stub, for example: https://github.com/carbon-design-system/ibm-dotcom-library/blob/2864816/packages/web-components/src/globals/services-store/actions/__tests__/localeAPI.test.ts#L82

Please don't hesitate to let me know if I'm missing something. Thanks!

**Comment by ccwilson-us-ibm-com** [comment]
A couple of thoughts...

Proposal 1 feels a little boiler platey.  So if we're mock 6 different items, we'll have to make 6 deep copies?  And each new item to be mocked will need to be added.

Proposal 2: That could work.  But on the subject of 'new devs', I don't know if that strategy reduces friction uniformly for new devs since the 'auto-mock' and 'mock reset' features of jest are not uniform across testing/mocking frameworks.  But my current PR isn't using anything that shouldn't already exist in any test/mock solution.

**Comment by asudoh** [comment]
Thank you for your additional thoughts @ccwilson-us-ibm-com!

> So if we're mock 6 different items, we'll have to make 6 deep copies?

No, you can keep one deep copy the mock module and the merge it back to the mock module for re-setting it to the original (that takes care of all properties you replace for specific tests).

> the 'auto-mock' and 'mock reset' features of jest are not uniform across testing/mocking frameworks.

Actually this comes from Jest out-of-the-box:

https://jestjs.io/docs/en/mock-functions#mocking-modules
https://jestjs.io/docs/en/configuration#resetmocks-boolean

**Comment by ccwilson-us-ibm-com** [comment]
Right.  They do come out of the box in Jest... but they don't in other testing/mocking frameworks... Jasmine, Sinon, etc.  As popular as it is right now, not every developer uses Jest.

**Comment by asudoh** [comment]
I see your point. Due to other reasons I have brought up, I'm for not introducing our own additional layer to our codebase for the sake of neutralizing the differences in testing libraries. Jest is popular and well documented, and utilizing its ecosystem and documentation looks more preferable for this problem space.

**Review by asudoh — COMMENTED** [review]
Thank you very much for the updates @ccwilson-us-ibm-com!

File: packages/services/src/services/Locale/__tests__/Locale.test.js [hunk 2]
```diff
@@ -88,47 +144,366 @@ describe('LocaleAPI', () => {
...
+      lc: 'testLC',
+    });
+
+    jest.advanceTimersByTime(100 * 51);
+
+    return listPromise.catch(() => {
+      resolvePromise({ data: response });
+      done();
+    });
+  }, 6000);
```
[the remark this answers is not in the document]
Comment (ccwilson-us-ibm-com) [reply packages/services/src/services/Locale/__tests__/Locale.test.js:299]: I've already applied you first note, and this is the current promise chain:
```
return listPromise.then(
      () => {
        throw new Error('`LocaleAPI.getList()` should have been rejected.');
      },
      () => {
        resolvePromise({ data: response });
      }
    );
```

If I take the 6 second timeout off of this test, it fails with a time out error.  If I add the 6 seconds back on, it succeeds.

**Comment by asudoh** [comment]
Going forward as-is for now as discussed with @jeffchew. @jeffchew Does the visual difference in footer in Percy look OK to you? Thanks!

**Comment by jeffchew** [comment]
> Going forward as-is for now as discussed with @jeffchew. @jeffchew Does the visual difference in footer in Percy look OK to you? Thanks!

Thanks @asudoh , I'm actually going to update the branch and rebuild and see what the new percy builds give us.

**Review by asudoh — APPROVED** [review]

**Pull request merged**